docs/architecture-overview.rst

```diff
@@ -0,0 +1,200 @@
+processes that process event queues.
+
+
+memcached
+---------
+
+memcached is used to cache database model
+objects. ``zerver/lib/cache.py`` and ``zerver/lib/cache_helpers.py``
+manage putting things into memcached, and its configuration is in in
+``puppet/zulip/files/memcached.conf``.
+
+Redis
+-----
+
+Redis is used for a few very short-term data stores, such as in the
+basis of ``zerver/lib/rate_limiter.py``, a per-user rate limiting
+scheme `example
+<http://blog.domaintools.com/2013/04/rate-limiting-with-redis/>`__).
+
+Redis is configured in ``zulip/puppet/zulip/files/redis`` and it's a
+pretty standard configuration except for the last line, which turns
+off persistence:
+
+::
+
+     # Zulip-specific configuration: disable saving to disk.
+     save ""
+
+memcached was used first and then we added Redis specifically to
+implement rate limiting. `We'd like to switch everything over to
+Redis.<https://github.com/zulip/zulip/issues/16>`__
+
+
+RabbitMQ
+--------
+
+RabbitMQ is a queueing system. Its config files live in
+``zulip/puppet/zulip/files/rabbitmq``. Initial configuration happens
+in ``zulip/scripts/setup/configure-rabbitmq``.
+
+We use RabbitMQ for queuing expensive work (e.g. sending emails
+triggered by a message, push notifications, some analytics, etc.) that
+require reliable delivery but which we don't want to do on the main
+thread. It's also used for communication between the application
+server and the Tornado push system.
+
+Two simple wrappers around ``pika`` (the Python RabbitMQ client) are
+in ``zulip/server/lib/queue.py``. There's an asynchronous client for
+use in Tornado and a more general client for use elsewhere.
+
+``zerver/lib/event_queue.py`` has helper functions for putting events
+into one queue or another. Most of the processes started by Supervise
+are queue processors that continually pull things out of a RabbitMQ
+queue and handle them.
+
+PostgreSQL
+----------
+
+PostgreSQL (also known as postgres) is the database that stores all
+persistent data, that is, data that's expected to live beyond a user's
+current session.
+
+In production, Postgres is installed with a default configuration. The
+directory that would contain configuration files
+(``puppet/zulip/files/postgresql``) has only a utility script and a
+custom list of stopwords used by a Postgresql extension.
+
+In a development environment, configuration of that postgresql
+extension is handled by ``tools/postgres-init-dev-db`` (invoked by
+``provision.py``). That file also manages setting up the development
+postgresql user.
+
+``provision.py`` also invokes ``tools/do-destroy-rebuild-database`` to
+create the actual database with its schema.
+
+Nagios
+------
+
+Nagios is an optional component used for notifications to the system administrator, e.g., in case of outages.
+
+``zulip/puppet/zulip/manifests/nagios.pp`` installs Nagios plugins
+from puppet/``zulip/files/nagios_plugins/``.
```

Comment:
We should be clear that this component is intended to install Nagios plugins intended to be run on a Nagios server; most of the Zulip Nagios plugins are intended to be run on the Zulip servers themselves, and are included with the relevant component of the Zulip server (e.g. puppet/zulip/manifests/postgres_common.pp installs a few under /usr/lib/nagios/plugins/zulip_postgres_common).